### PR TITLE
Add easy access to `state_dict` in Lite module wrapper

### LINKED ISSUE #14571: Simplify state dict access on model wrapper in Lite
### Motivation

Currently, the user has to be aware of whether their model is wrapped or not. To save a checkpoint, they need to call
model.module to access the state dict. Example code:

```py
import torch

from pytorch_lightning.lite import LightningLite


class Lite(LightningLite):
    def run(self):
        model = torch.nn.Linear(2, 2)
        model = self.setup(model)
        print(model.state_dict().keys())

        # Current:
        # _forward_module.module.weight, _forward_module.module.bias
        # _original_module.weight, _original_module.bias

        # Expected:
        # weight, bias

        # Note: .module gives the expected keys
        print(model.module.state_dict().keys())


if __name__ == "__main__":
    Lite(accelerator="cpu", strategy="ddp", devices=2).run()

```

As you can see, not only do we get the prefixes on the keys, but also duplicate weight references, because internally we save a reference to the original module in addition to the ddp wrapper. 

This could also be seen as unintended behavior, and be fixed in a bug release. 

### Pitch

Implement the `state_dict` and `load_state_dict` methods on the wrapper, and redirect the call to the original module. 

### Alternatives

Keep as is, but make it clear in the docs that user needs to call `model.module` to access the original model.

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging PyTorch Lightning, Transformers, and Hydra.

cc @borda @justusschock @carmocca @awaelchli

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #14571

## Breaking change

This is a breaking change for anyone who had to work-around the saving and loading with `model.module`, or mangling with the raw state-dict keys of `model.state_dict()`.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: src/pytorch_lightning/lite/wrappers.py [hunk 1]
```diff
@@ -115,6 +117,26 @@ def _convert_float_tensor(t: Tensor) -> Tensor:
...
+
+    @overload
+    def state_dict(self, *, prefix: str = ..., keep_vars: bool = ...) -> T_destination:
+        ...
+
+    def state_dict(
+        self, *args: Any, destination: Optional[T_destination] = None, prefix: str = "", keep_vars: bool = False
+    ) -> Optional[T_destination]:
+        return self._original_module.state_dict(
+            destination=destination,  # type: ignore[type-var]
```
Comment (awaelchli) [inline src/pytorch_lightning/lite/wrappers.py:132]: The type hints in nn.Module are unfortunately a bit messed up

File: src/pytorch_lightning/lite/wrappers.py [hunk 2]
```diff
@@ -115,6 +117,26 @@ def _convert_float_tensor(t: Tensor) -> Tensor:
...
+    def state_dict(self, *, destination: T_destination, prefix: str = ..., keep_vars: bool = ...) -> T_destination:
+        ...
+
+    @overload
+    def state_dict(self, *, prefix: str = ..., keep_vars: bool = ...) -> T_destination:
+        ...
+
+    def state_dict(
+        self, *args: Any, destination: Optional[T_destination] = None, prefix: str = "", keep_vars: bool = False
+    ) -> Optional[T_destination]:
```
Comment (carmocca) [inline src/pytorch_lightning/lite/wrappers.py:130]: Why optional?

File: src/pytorch_lightning/lite/wrappers.py [hunk 3]
```diff
@@ -115,6 +117,26 @@ def _convert_float_tensor(t: Tensor) -> Tensor:
...
+        return self._original_module.state_dict(
+            destination=destination,  # type: ignore[type-var]
+            prefix=prefix,
+            keep_vars=keep_vars,
+        )
+
+    def load_state_dict(self, state_dict: Mapping[str, Any], strict: bool = True) -> Any:
```
Comment (carmocca) [inline src/pytorch_lightning/lite/wrappers.py:137]: Doesnt this always return None?

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/lite/wrappers.py (hunk 2 above)
Comment (awaelchli) [reply src/pytorch_lightning/lite/wrappers.py:130]: If I don't, mypy infers it can be optional. I'm just trying to override torch's method, which doesn't use mypy. If you have a better suggestion, please modify it.

File: src/pytorch_lightning/lite/wrappers.py (hunk 3 above)
Comment (awaelchli) [reply src/pytorch_lightning/lite/wrappers.py:137]: See nn.Module implementation, which has no types (torch 1.12). I'm adding what mypy infers. Please edit directly if you have another solution.

File: src/pytorch_lightning/lite/wrappers.py (hunk 2 above)
Comment (justusschock) [reply src/pytorch_lightning/lite/wrappers.py:130]: In torch they just don't provide any typing (including return types) as the types should be clear from overloads: https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L1374

**Review by justusschock — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: src/pytorch_lightning/lite/wrappers.py (hunk 3 above)
Comment (rohitgr7) [reply src/pytorch_lightning/lite/wrappers.py:137]: it returns a NamedTuple with missing keys

**Review by rohitgr7 — APPROVED** [review]

**Comment by Borda** [comment]
@awaelchli, mind checking/resolving the remaining conversations? :otter:

File: src/pytorch_lightning/lite/wrappers.py (hunk 2 above)
Comment (carmocca) [reply src/pytorch_lightning/lite/wrappers.py:131]: Okay. Not sure how to fix this so resolving

File: src/pytorch_lightning/lite/wrappers.py [hunk 4]
```diff
@@ -115,6 +118,26 @@ def _convert_float_tensor(t: Tensor) -> Tensor:
...
+    @overload
+    def state_dict(self, *, destination: T_destination, prefix: str = ..., keep_vars: bool = ...) -> T_destination:
+        ...
+
+    @overload
+    def state_dict(self, *, prefix: str = ..., keep_vars: bool = ...) -> T_destination:
+        ...
+
+    def state_dict(
+        self, *args: Any, destination: Optional[T_destination] = None, prefix: str = "", keep_vars: bool = False
```
Comment (carmocca) [inline src/pytorch_lightning/lite/wrappers.py:130]: I would remove `*args` since you aren't passing them. They are deprecated upstream anyways

```suggestion
        self, destination: Optional[T_destination] = None, prefix: str = "", keep_vars: bool = False
```

File: src/pytorch_lightning/lite/wrappers.py [hunk 5]
```diff
@@ -115,6 +118,26 @@ def _convert_float_tensor(t: Tensor) -> Tensor:
         output = apply_to_collection(output, function=_convert_float_tensor, dtype=Tensor)
         return output
 
+    @overload
+    def state_dict(self, *, destination: T_destination, prefix: str = ..., keep_vars: bool = ...) -> T_destination:
+        ...
+
+    @overload
+    def state_dict(self, *, prefix: str = ..., keep_vars: bool = ...) -> T_destination:
+        ...
```
Comment (carmocca) [inline src/pytorch_lightning/lite/wrappers.py:127]: Why do we define these overloads? I tried removing them and doesn't seem like anything changes

Comment (justusschock) [reply src/pytorch_lightning/lite/wrappers.py:127]: this does help a lot with IDE completion and should enable us to remove the typing of the method's implementation at all as this already defines typing.

Comment (awaelchli) [reply src/pytorch_lightning/lite/wrappers.py:127]: I'm sorry, I can't remove the types @justusschock, our mypy says: Missing type annotation for function etc.

> Why do we define these overloads? I tried removing them and doesn't seem like anything changes

@carmocca I get this: 

```
src/pytorch_lightning/lite/wrappers.py:121: error: Signature of "state_dict" incompatible with supertype "Module"  [override]
src/pytorch_lightning/lite/wrappers.py:121: note:      Superclass:
src/pytorch_lightning/lite/wrappers.py:121: note:          @overload
src/pytorch_lightning/lite/wrappers.py:121: note:          def [T_destination <: Dict[str, Any]] state_dict(self, *, destination: T_destination, prefix: str = ..., keep_vars: bool = ...) -> T_destination
src/pytorch_lightning/lite/wrappers.py:121: note:          @overload
src/pytorch_lightning/lite/wrappers.py:121: note:          def state_dict(self, *, prefix: str = ..., keep_vars: bool = ...) -> Dict[str, Any]
src/pytorch_lightning/lite/wrappers.py:121: note:      Subclass:
src/pytorch_lightning/lite/wrappers.py:121: note:          def [T_destination <: Dict[str, Any]] state_dict(self, destination: Optional[T_destination] = ..., prefix: str = ..., keep_vars: bool = ...) -> Optional[T_destination]

```

@carmocca @justusschock feel free to push any changes regarding typing, as long as it is valid. I don't care too much about the typing details here.
